pkg/dedup/chunk_iter.go

```diff
@@ -0,0 +1,357 @@
+		},
+	}
+}
+
+func (o *overlappingMerger) addChunk(chk chunks.Meta) {
+	switch chk.Chunk.Encoding() {
+	case chunkenc.EncXOR:
+		o.xorIterators = append(o.xorIterators, chk.Chunk.Iterator(nil))
+	case downsample.ChunkEncAggr:
+		aggrChk := chk.Chunk.(*downsample.AggrChunk)
+		for i := downsample.AggrCount; i <= downsample.AggrCounter; i++ {
+			if c, err := aggrChk.Get(i); err == nil {
+				o.aggrIterators[i] = append(o.aggrIterators[i], c.Iterator(nil))
+			}
+		}
+	}
+}
+
+func (o *overlappingMerger) empty() bool {
+	// OverlappingMerger only contains either xor chunk or aggr chunk.
+	// If xor chunks are present then we don't need to check aggr chunks.
+	if len(o.xorIterators) > 0 {
+		return false
+	}
+	return len(o.aggrIterators[downsample.AggrCount]) == 0
+}
+
+// Return a chunk iterator based on the encoding of base chunk.
+func (o *overlappingMerger) iterator(baseChk chunks.Meta) chunks.Iterator {
+	var it chunkenc.Iterator
+	switch baseChk.Chunk.Encoding() {
+	case chunkenc.EncXOR:
+		// If XOR encoding, we need to deduplicate the samples and re-encode them to chunks.
+		return storage.NewSeriesToChunkEncoder(&storage.SeriesEntry{
+			SampleIteratorFn: func() chunkenc.Iterator {
+				it = baseChk.Chunk.Iterator(nil)
+				for _, i := range o.xorIterators {
+					it = o.samplesMergeFunc(it, i)
+				}
+				return it
+			}}).Iterator()
+
+	case downsample.ChunkEncAggr:
+		// If Aggr encoding, each aggregated chunks need to be expanded and deduplicated,
+		// then re-encoded into Aggr chunks.
+		aggrChk := baseChk.Chunk.(*downsample.AggrChunk)
+		samplesIter := [5]chunkenc.Iterator{}
+		for i := downsample.AggrCount; i <= downsample.AggrCounter; i++ {
+			if c, err := aggrChk.Get(i); err == nil {
+				o.aggrIterators[i] = append(o.aggrIterators[i], c.Iterator(nil))
+			}
+
+			if len(o.aggrIterators[i]) > 0 {
+				for _, j := range o.aggrIterators[i][1:] {
+					o.aggrIterators[i][0] = o.samplesMergeFunc(o.aggrIterators[i][0], j)
+				}
+				samplesIter[i] = o.aggrIterators[i][0]
+			} else {
+				samplesIter[i] = nil
+			}
+		}
+
+		return newAggrChunkIterator(samplesIter)
+	}
+
+	// Impossible for now.
+	return nil
+}
+
+type aggrChunkIterator struct {
+	iters        [5]chunkenc.Iterator
+	curr         chunks.Meta
+	countChkIter chunks.Iterator
+
+	err error
+}
+
+func newAggrChunkIterator(iters [5]chunkenc.Iterator) chunks.Iterator {
+	return &aggrChunkIterator{
+		iters: iters,
+		countChkIter: storage.NewSeriesToChunkEncoder(&storage.SeriesEntry{
+			SampleIteratorFn: func() chunkenc.Iterator {
+				return iters[downsample.AggrCount]
+			},
+		}).Iterator(),
+	}
+}
+
+func (a *aggrChunkIterator) Next() bool {
+	if !a.countChkIter.Next() {
+		return false
+	}
+
+	countChk := a.countChkIter.At()
+	mint := countChk.MinTime
+	maxt := countChk.MaxTime
+
+	var (
+		chks [5]chunkenc.Chunk
+		chk  *chunks.Meta
+		err  error
+	)
+
+	chks[downsample.AggrCount] = countChk.Chunk
+	chk, err = a.toChunk(downsample.AggrSum, mint, maxt)
```

Comment:
This block here probably needs a for-loop instead of copy/paste :smiley:

Reply:
Done in https://github.com/thanos-io/thanos/pull/4239/commits/d0cba248a2914669389d31d06e517327df272b6d